File: app/src/androidTest/java/com/lunesu/pengchauferry/FerryDaoTest.kt
package com.lunesu.pengchauferry

import androidx.test.ext.junit.runners.AndroidJUnit4
import org.joda.time.Duration
import org.joda.time.LocalTime
import org.junit.After
import org.junit.Assert.assertEquals
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class FerryDaoTest {
    private val db = DbOpenHelper(null)
    private val ferry = Ferry(LocalTime.MIDNIGHT, FerryPier.MuiWo, FerryPier.PengChau, Duration.ZERO, FerryDay.MondayToSaturday, "1.2", FerryPier.HeiLingChau)

    @After
    fun tearDown() {
        db.close()
    }

    @Test
    fun testQuery() {
        val dao = FerryDao(db)
        assertEquals(emptyList<Ferry>(), dao.query(FerryPier.MuiWo, FerryDay.Holiday))
    }

    @Test
    fun testInsert() {
        val dao = FerryDao(db)
        dao.insert(ferry)
        assertEquals(listOf(ferry), dao.query(ferry.from, FerryDay.Monday))
    }

    @Test
    fun testQueryFilter() {
        val dao = FerryDao(db)
        dao.insert(ferry)
        assertEquals(emptyList<Ferry>(), dao.query(ferry.from, FerryDay.Holiday))
        assertEquals(emptyList<Ferry>(), dao.query(FerryPier.ChiMaWan, FerryDay.Monday))
    }

    @Test
    fun testInsertTwice() {
        val dao = FerryDao(db)
        dao.insert(ferry)
        dao.insert(ferry)
    }

    @Test
    fun testDelete() {
        val dao = FerryDao(db)
        dao.insert(ferry)
        dao.delete(ferry.from, ferry.to)
        assertEquals(emptyList<Ferry>(), dao.query(ferry.from, FerryDay.Monday))
    }

    @Test
    fun testSave() {
        val dao = FerryDao(db)
        dao.save(listOf(ferry), ferry.from, ferry.to)
        assertEquals(listOf(ferry), dao.query(FerryPier.MuiWo, FerryDay.Monday))
    }
}

File: iOS/PengChauFerryTests/FerryDaoTest.swift
//
//  FerryDaoTest.swift
//  PengChauFerryTests
//
//  Created by Lionello Lunesu on 2/8/2020.
//  Copyright © 2020 Lionello Lunesu. All rights reserved.
//

import XCTest
@testable import PengChauFerry

class FerryDaoTest: XCTestCase {

    private let ferry = Ferry(time: .MIDNIGHT, from: .MuiWo, to: .PengChau, dur: 0, days: .MondayToSaturday, fare: "1.2", via: .HeiLingChau)

    func testQuery() {
        let dao = FerryDao()
        XCTAssertEqual([], dao.query(from: .MuiWo, dow: .Holiday))
    }

    func testInsert() {
        let dao = FerryDao()
        dao.insert(ferry: ferry)
        XCTAssertEqual([ferry], dao.query(from: ferry.from, dow: .Monday))
    }

    func testQueryFilter() {
        let dao = FerryDao()
        dao.insert(ferry: ferry)
        XCTAssertEqual([], dao.query(from: ferry.from, dow: .Holiday))
        XCTAssertEqual([], dao.query(from: .ChiMaWan, dow: .Monday))
    }

    func testInsertTwice() {
        let dao = FerryDao()
        dao.insert(ferry: ferry)
        dao.insert(ferry: ferry)
    }

    func testDelete() {
        let dao = FerryDao()
        dao.insert(ferry: ferry)
        dao.delete(piers: ferry.from, ferry.to)
        XCTAssertEqual([], dao.query(from: ferry.from, dow: .Monday))
    }

    func testSave() {
        let dao = FerryDao()
        dao.save(result: [ferry], piers: ferry.from, ferry.to)
        XCTAssertEqual([ferry], dao.query(from: .MuiWo, dow: .Monday))
    }

    func testGet() {
        let dao = FerryDao()
        dao.insert(ferry: ferry)
        XCTAssertEqual(ferry, dao.get(from: ferry.from, to: ferry.to, dow: .Monday, time: LocalTime.MIDNIGHT))
        XCTAssertNil(dao.get(from: ferry.from, to: ferry.to, dow: .Sunday, time: LocalTime.MIDNIGHT))
        XCTAssertNil(dao.get(from: ferry.from, to: ferry.from, dow: .Monday, time: LocalTime.MIDNIGHT))
        XCTAssertNil(dao.get(from: ferry.from, to: ferry.from, dow: .Monday, time: LocalTime(secs: 671)))
        XCTAssertNil(dao.get(from: ferry.to, to: ferry.to, dow: .Monday, time: LocalTime.MIDNIGHT))
    }

}
